### PR TITLE
MESH-2096/allow-sc-deploy-as-child-id

### PR DESCRIPTION
**Pull request by HenriqueNogara**
Allows a contract to be deployed as a primary key of a child identity.

### new features

- Adds the following extrisincs: `instantiate_with_code_as_primary_key`, `instantiate_with_hash_as_primary_key`;

### modified logic

- `on_instantiate_transfer` if `contract_did` is different from the caller's did, also check if its parent match the caller's identity;

### CODE REVIEW

File: pallets/contracts/src/lib.rs [hunk 1]
```diff
@@ -333,7 +338,7 @@ decl_module! {
         /// - All the errors in `pallet_contracts::Call::instantiate_with_code` can also happen here.
         /// - CDD/Permissions are checked, unlike in `pallet_contracts`.
         /// - Errors that arise when adding a new secondary key can also occur here.
-        #[weight = Module::<T>::weight_instantiate_with_code(&code, &salt, &perms).saturating_add(*gas_limit)]
+        #[weight = Module::<T>::weight_instantiate_with_code(&code, &salt, &Some(perms.clone())).saturating_add(*gas_limit)]
```
Comment (Neopallium) [inline pallets/contracts/src/lib.rs]:
```suggestion
        #[weight = Module::<T>::weight_instantiate_with_code(&code, &salt, Some(&perms)).saturating_add(*gas_limit)]
```
Don't need a `clone` here.  Option can wrap a reference.

File: pallets/contracts/src/lib.rs [hunk 2]
```diff
@@ -438,27 +535,40 @@ where
...
-    /// Computes weight of `instantiate_with_code(code, salt, perms)`.
-    fn weight_instantiate_with_code(code: &[u8], salt: &[u8], perms: &Permissions) -> Weight {
-        <T as Config>::WeightInfo::instantiate_with_code_bytes(&code, &salt).saturating_add(
-            <T as IdentityConfig>::WeightInfo::permissions_cost_perms(perms),
-        )
+    /// Computes the weight of `instantiate_with_code(code, salt, perms)`.
+    fn weight_instantiate_with_code(
+        code: &[u8],
+        salt: &[u8],
+        perms: &Option<Permissions>,
```
Comment (Neopallium) [inline pallets/contracts/src/lib.rs]:
```suggestion
        perms: Option<&Permissions>,
```

File: pallets/contracts/src/lib.rs [hunk 3]
```diff
@@ -483,14 +594,21 @@ where
...
         )
     }
 
-    /// Computes weight of `instantiate_with_hash(code, salt, perms)`.
-    fn weight_instantiate_with_hash(salt: &[u8], perms: &Permissions) -> Weight {
-        <T as Config>::WeightInfo::instantiate_with_hash_bytes(&salt).saturating_add(
-            <T as IdentityConfig>::WeightInfo::permissions_cost_perms(perms),
-        )
+    /// Computes weight of `instantiate_with_hash(salt, perms)`.
+    fn weight_instantiate_with_hash(salt: &[u8], perms: &Option<Permissions>) -> Weight {
```
Comment (Neopallium) [inline pallets/contracts/src/lib.rs]:
```suggestion
    fn weight_instantiate_with_hash(salt: &[u8], perms: Option<&Permissions>) -> Weight {
```

**Review by Neopallium — CHANGES_REQUESTED** [review]
Minor extra clones.  Otherwise looks good.

**Review by Neopallium — APPROVED** [review]

**Review by adamdossa — APPROVED** [review]

**Pull request merged**